app/assets/javascripts/discourse/app/models/topic-tracking-state.js

```diff
@@ -577,21 +553,279 @@ const TopicTrackingState = EmberObject.extend({
+      //
+      // TODO (martin) add the list limit
+      if (!topic.prevent_sync) {
+        this.removeTopic(topic.id);
+      }
+      return;
+    }
+
+    newState.highest_post_number = topic.highest_post_number;
+    if (topic.category) {
+      newState.category_id = topic.category.id;
+    }
+
+    if (topic.tags) {
+      newState.tags = topic.tags;
+    }
+
+    this.modifyState(topic.id, newState);
+  },
 
-    // I am taking some shortcuts here to avoid 500 gets for a large list
-    if (data) {
-      data.forEach((topic) => {
-        states["t" + topic.topic_id] = topic;
+  // this stops sync of tracking state when list is filtered, in the past this
+  // would cause the tracking state to become inconsistent.
+  _shouldCompensateState(list, filter, queryParams) {
+    let shouldCompensate =
+      (filter === "new" || filter === "unread") && !list.more_topics_url;
+
+    if (shouldCompensate && queryParams) {
+      Object.keys(queryParams).forEach((k) => {
+        if (k !== "ascending" && k !== "order") {
+          shouldCompensate = false;
+        }
       });
     }
+
+    return shouldCompensate;
+  },
+
+  // any state that is not in the provided list must be updated
+  // based on the filter selected so we do not have any incorrect
+  // state in the list
+  _correctMissingState(list, filter) {
+    const ids = {};
+    list.topics.forEach((topic) => (ids[this._stateKey(topic.id)] = true));
+
+    Object.keys(this.states).forEach((topicKey) => {
+      // if the topic is already in the list then there is
+      // no compensation needed; we already have latest state
+      // from the backend
+      if (ids[topicKey]) {
+        return;
+      }
+
+      const newState = { ...this.findState(topicKey) };
+      if (filter === "unread" && isUnread(newState)) {
+        // pretend read. if unread, the highest_post_number will be greater
+        // than the last_read_post_number
+        newState.last_read_post_number = newState.highest_post_number;
+      }
+
+      if (filter === "new" && isNew(newState)) {
+        // pretend not new. if the topic is new, then last_read_post_number
+        // will be null.
+        newState.last_read_post_number = 1;
+      }
+
+      this.modifyState(topicKey, newState);
+    });
+  },
+
+  // processes the data sent via messageBus, called by establishChannels
+  _processChannelPayload(data) {
+    if (["muted", "unmuted"].includes(data.message_type)) {
+      this.trackMutedOrUnmutedTopic(data);
+      return;
+    }
+
+    this.pruneOldMutedAndUnmutedTopics();
+
+    if (this.isMutedTopic(data.topic_id)) {
+      return;
+    }
+
+    if (
+      this.siteSettings.mute_all_categories_by_default &&
+      !this.isUnmutedTopic(data.topic_id)
+    ) {
+      return;
+    }
+
+    // TODO(martin) - The /delete channel handles these messages and doesn't
+    // even send them for process, so this can be removed?
+    if (data.message_type === "delete") {
```

Comment:
I am not really sure if we want to set topic.deleted like we do here:

https://github.com/discourse/discourse/pull/12555/files#diff-0ede46eb4b3a984cbf5fedf65c61d9076a78fb568582ffd245254099534a00cbR79-R81

Or if we just want to remove the topic completely. This code on 720 is not currently being called.